File: programmers/2023-kakao-blind-recruitment/개인정보-수집-유효기간/main.js
function solution(today, terms, privacies) {
  var answer = [];
  const todayDate = new Date(today);
  const termPair = new Map();
  Promise.all(
    terms.map((term) => {
      const termArray = term.split(' ');
      termPair.set(termArray[0], Number(termArray[1]));
    })
  );
  Promise.all(
    privacies.map((privacy, index) => {
      const privacyArray = privacy.split(' ');
      const date = new Date(privacyArray[0]);
      date.setMonth(date.getMonth() + termPair.get(privacyArray[1]));
      if (date <= todayDate) {
        answer.push(index + 1);
      }
    })
  );
  return answer;
}

// best solution
function solution(today, terms, privacies) {
  var answer = [];
  var [year, month, date] = today.split('.').map(Number);
  var todates = year * 12 * 28 + month * 28 + date;
  var t = {};
  terms.forEach((e) => {
    let [a, b] = e.split(' ');
    t[a] = Number(b);
  });
  privacies.forEach((e, i) => {
    var [day, term] = e.split(' ');
    day = day.split('.').map(Number);
    var dates = day[0] * 12 * 28 + day[1] * 28 + day[2] + t[term] * 28;
    if (dates <= todates) answer.push(i + 1);
  });
  return answer;
}

File: programmers/2023-kakao-blind-recruitment/개인정보-수집-유효기간/main.py
from datetime import date, datetime
from dateutil.relativedelta import relativedelta


def solution(today, terms, privacies):
    answer = []
    termMap = {}
    todayArray = today.split('.')
    todayDate = date(int(todayArray[0]), int(
        todayArray[1]), int(todayArray[2]))
    for term in terms:
        termArray = term.split(' ')
        termMap[termArray[0]] = termArray[1]

    for i in range(len(privacies)):
        privacyArray = privacies[i].split(' ')
        dateArray = privacyArray[0].split('.')
        date_after_term = date(
            int(dateArray[0]), int(dateArray[1]), int(dateArray[2])) + relativedelta(months=int(termMap[privacyArray[1]]))
        if date_after_term <= todayDate:
            answer.append(i+1)

    return answer


# best solution
def to_days(date):
    year, month, day = map(int, date.split("."))
    return year * 28 * 12 + month * 28 + day


def solution(today, terms, privacies):
    months = {v[0]: int(v[2:]) * 28 for v in terms}
    today = to_days(today)
    expire = [
        i + 1 for i, privacy in enumerate(privacies)
        if to_days(privacy[:-2]) + months[privacy[-1]] <= today
    ]
    return expire
